crates/sui-framework/packages/sui-system/sources/validator_set.move

```diff
@@ -445,51 +480,69 @@ public(package) fun advance_epoch(
         ctx
     );
-
-    self.total_stake = calculate_total_stakes(&self.active_validators);
-
-    voting_power::set_voting_power(&mut self.active_validators);
+    self.total_stake = new_total_stake;
+    voting_power::set_voting_power(&mut self.active_validators, new_total_stake);
 
     // At this point, self.active_validators are updated for next epoch.
     // Now we process the staged validator metadata.
     effectuate_staged_metadata(self);
 }
 
-fun update_and_process_low_stake_departures(
+/// This function does the following:
+/// - removes validators from `at_risk` group if their voting power is above the LOW threshold
+/// - increments the number of epochs a validator has been below the LOW threshold but above the
+///     VERY LOW threshold
+/// - removes validators from the active set if they have been below the LOW threshold for more than
+///     `low_stake_grace_period` epochs
+/// - removes validators from the active set immediately if they are below the VERY LOW threshold
+/// - activates pending validators if they have sufficient voting power
+fun update_validator_positions_and_calculate_total_stake(
     self: &mut ValidatorSet,
-    low_stake_threshold: u64,
-    very_low_stake_threshold: u64,
     low_stake_grace_period: u64,
     validator_report_records: &mut VecMap<address, VecSet<address>>,
     ctx: &mut TxContext
-) {
+): u64 {
+    // take all pending validators out of the tablevec and put them in a local vector
+    let pending_active_validators = vector::tabulate!(
+        self.pending_active_validators.length(),
+        |_| self.pending_active_validators.pop_back()
+    );
+
+    // Note: we count the total stake of pending validators as well!
+    let pending_total_stake = calculate_total_stakes(&pending_active_validators);
+    let initial_total_stake = calculate_total_stakes(&self.active_validators) + pending_total_stake;
+    let (min_joining_voting_power_threshold, low_voting_power_threshold, very_low_voting_power_threshold) = self.get_voting_power_thresholds(ctx);
     // Iterate through all the active validators, record their low stake status, and kick them out if the condition is met.
+    let mut total_removed_stake = 0; // amount of stake to remove due to departed_validators
     let mut i = self.active_validators.length();
     while (i > 0) {
         i = i - 1;
         let validator_ref = &self.active_validators[i];
         let validator_address = validator_ref.sui_address();
-        let stake = validator_ref.total_stake_amount();
-        if (stake >= low_stake_threshold) {
+        let validator_stake = validator_ref.total_stake_amount();
+
+        // calculate the voting power for this validator in the next epoch if no validators are removed
+        // if one of more low stake validators are removed, it's possible this validator will have higher voting power--that's ok.
+        let voting_power = voting_power::derive_raw_voting_power(validator_stake, initial_total_stake);
```

Comment:
just so i undersatnd - this is actually the voting power of the validator assuming it plus all other pending validators are added, right?

is it possible we could get into a situation where there are so many pending validators none of them can have enough voting power if all are added, even though a subset of them would be okay?